cuda/components/atomic.cuh

```diff
@@ -39,11 +39,73 @@ namespace kernels {
 namespace cuda {
 
 
+namespace detail {
+
+
+template <typename ValueType, typename = void>
+struct atomic_add {
+    __forceinline__ __device__ void operator()(ValueType *, ValueType)
+    {
+        GKO_ASSERT(false);
+        // TODO: add proper implementation of generic atomic add
+    }
+};
+
+
+template <typename ResultType, typename ValueType>
+__forceinline__ __device__ ResultType as(ValueType val)
+{
+    return reinterpret_cast<ResultType &>(val);
+}
+
+
+#define GKO_BIND_ATOMIC_ADD_STRUCTURE(CONVERTER_TYPE)                     \
+    template <typename ValueType>                                         \
+    struct atomic_add<ValueType,                                          \
+                      gko::xstd::enable_if_t<(sizeof(ValueType) ==        \
+                                              sizeof(CONVERTER_TYPE))>> { \
+        __forceinline__ __device__ void operator()(                       \
```

Comment:
You mean I use the specific function name like `atomic_add` not overload `operator()` and I can use like atomic_helper<T>::atomic_add().
Is this correct?
Or do you mean some the static function like `static void func(...)` approach?

Reply:
Yes, exactly as your first version (with `atomic_add` being a static function). In my opinion, that makes it a lot more readable.

Reply:
I think I understand your suggestion more clear.
If I would like to use `atomic_helper<T>::atomic_add` directly without constructing `atomic_helper`, I need to declare the static function.
Otherwise, I will get the error `a nonstatic member reference must be relative to a specific object` from compiler.

Reply:
Yes, this is exactly what I meant, thanks!
I probably should have provided a snippet how it would be called (exactly what you wrote: `atomic_helper<T>::atomic_add(...)`) in my first comment to make it more clear.